kernel/hashlib.h

```diff
@@ -356,6 +358,29 @@ template<typename K, int offset = 0, typename OPS = hash_ops<K>> class idict;
 template<typename K, typename OPS = hash_ops<K>> class pool;
 template<typename K, typename OPS = hash_ops<K>> class mfp;
 
+// Computes the hash value of an unordered set of elements.
+// See https://www.preprints.org/manuscript/201710.0192/v1/download.
+// This is the Sum(4) algorithm from that paper, which has good collision resistance,
+// much better than Sum(1) or Xor(1) (and somewhat better than Xor(4)).
+class commutative_hash {
+public:
+	commutative_hash() {
+		buckets.fill(0);
+	}
+	void eat(Hasher h) {
+		Hasher::hash_t v = h.yield();
+		size_t index = v & (buckets.size() - 1);
+		buckets[index] += v;
+	}
```

Comment:
Thanks, that is an interesting paper.  A minor point is that paper describes Sum(4) as:

Sum(4): the radix sort-inspired method where the bottom 2 bits of each element hash selects an accumulator, the **remaining bits** are summed into that accumulator, and at the end the accumulators and set size are combined.

Here you sum all the bits not the remaining bits.  I doesn't seem to matter empirically.